packages/scoutgame/src/protocol/resolveTokenOwnership.ts

```diff
@@ -0,0 +1,135 @@
+import { log } from '@charmverse/core/log';
+import { getPublicClient } from '@packages/blockchain/getPublicClient';
+import type { Address } from 'viem';
+import { baseSepolia, optimism } from 'viem/chains';
+
+import { getTransferSingleWithBatchMerged } from '../builderNfts/accounting/getTransferSingleWithBatchMerged';
+import type { ISOWeek } from '../dates';
+
+async function getLastBlockOfWeek({ week, chainId }: { week: string; chainId: number }): Promise<number> {
+  // Step 1: Parse the ISO week (e.g., '2024-W50') and calculate the end of the week (Sunday 23:59:59 UTC)
+  const lastSundayUTC = getEndDateOfISOWeek(week);
+  const targetTimestamp = Math.floor(lastSundayUTC.getTime() / 1000); // Convert to Unix timestamp (seconds)
+
+  log.info(`Last Sunday UTC at 23:59:59 timestamp: ${targetTimestamp} (Date: ${lastSundayUTC.toUTCString()})`);
+
+  // Step 2: Get the public client for the chain
+  const client = getPublicClient(chainId);
+
+  // Get the current latest block
+  const blockNumber = await client.getBlockNumber();
+  const latestBlock = await client.getBlock({ blockNumber });
+
+  // Step 3: Binary search for the block closest to the target timestamp
+  let low = 0;
+  let high = Number(blockNumber);
+  let resultBlock = latestBlock;
+
+  while (low <= high) {
+    const mid = Math.floor((low + high) / 2);
+    const block = await client.getBlock({ blockNumber: BigInt(mid) });
+
+    if (block.timestamp < targetTimestamp) {
+      low = mid + 1;
+    } else if (block.timestamp > targetTimestamp) {
+      high = mid - 1;
+    } else {
+      resultBlock = block;
+      break;
+    }
+
+    resultBlock = block; // Store the latest "best match" block
+  }
+
+  log.info(
+    `Found block at ${resultBlock.timestamp} (Date: ${new Date(Number(resultBlock.timestamp) * 1000).toUTCString()})`
+  );
+  log.info(`Block Number: ${resultBlock.number}`);
+
+  return Number(resultBlock.number);
+}
+
+/**
+ * Calculates the end date (Sunday 23:59:59 UTC) of the given ISO week (YYYY-Www).
+ * @param week - The ISO week in the format 'YYYY-Www' (e.g., '2024-W50').
+ * @returns A Date object set to Sunday 23:59:59 UTC of that ISO week.
+ */
+function getEndDateOfISOWeek(week: string): Date {
```

Comment:
This is crazy complicated. Please just use  getWeekStartEnd from dates please

Reply:
Fixed